Documentation/specs/corehost.md

```diff
@@ -0,0 +1,112 @@
+
+This index is loaded when needed during the Resolution Process (see below).
+
+### Dependencies File
+
+The `.deps` file is a file located alongside the native host. It is stored in a CSV format, with header, of the following format:
+
+```
+LibraryType,LibraryName,LibraryVersion,LibraryHash,AssetType,AssetName,AssetRelativePath
+"Package","Microsoft.CodeAnalysis.Common","1.2.0-beta-20151117-04","sha512-4O3f1iRswbWDz70AGTk0tBA1ivslJUuq37FSx4d0+ZuoBFxNO4oY7ReJgndOJYefaxeTMsCVHSVHJMYOJJphtQ==","runtime","Microsoft.CodeAnalysis","lib\\portable-net45+win8\\Microsoft.CodeAnalysis.dll",
+```
+
+Fields:
+
+* `LibraryType` -> The type of Library that provides the Asset. Currently always `Package` (other Library types include Projects and Framework Assemblies but these aren't relevant to `corehost`)
+* `LibraryName` -> The name of the Library that provides the Asset. For a Package, this is the Package ID
+* `LibraryVersion` -> The version of the Library that provides the Asset. For a Package, this is the Package Version
+* `LibraryHash` -> The hash of the Library that provides the Asset. For a Package, this is a hash of the `.nupkg` file. The format is `[algorithm]-[base64-encoded hash]` where `[algorithm]` must always be `sha512` (in the current version)
+* `AssetType` -> The type of the Asset within the Library. `corehost` recognizes two types: `runtime` and `native`, representing Managed Assemblies and Native Libraries respectively.
+* `AssetName` -> The base name of the Asset. This is generally just the file name, without the extension (also removing "sub-extensions" like `ni` for native images of managed assemblies)
+* `AssetRelativePath` -> The relative path to the Asset file within the root of the Library.
+
+A few notes about the format:
+
+* The current code does not include the header, this is being changed.
+* All fields MUST be quoted (for simplicity), even if they don't contain a ','
+
+### Files in the application folder
+
+Any file with the suffix `.dll` in the same folder as the managed application being loaded (the "Application Base") will be considered a viable assembly during the resolution process. The host **assumes** that the assembly's short name is the same as the file name with the `.dll` suffix removed (yes, this is not technically required by the CLR, but we assume it for use with this host).
+
+### Files from package caches
+
+Only assemblies listed in the dependencies file can be resolved from a package cache. To resolve those assemblies, two environment variables are used:
+
+* `DOTNET_PACKAGES` - The primary package cache. If not set, defaults to `$HOME/.nuget/packages` on Unix or `%LOCALAPPDATA%\NuGet\Packages` (TBD) on Windows. **NOTE**: Currently the host uses different folders as we are still coordinating with NuGet to get the directories right (there are compat considerations). Currently we always use `$HOME/.dnx/packages`(Unix)/`%USERPROFILE%\.dnx\packages`(Win).
```

Comment:
Generally we'd expect these to be set globally and overridden in individual contexts only when necessary. For example, a shared hoster would set `DOTNET_HOME` to their shared installation of the runtime, and could allow apps to override that runtime either by providing their own value or by publishing a runtime with their app (app-local runtime beats `DOTNET_HOME`). Similarly for `DOTNET_PACKAGES_CACHE`, etc.

There is an open discussion around apps taking dependencies on specific runtime versions. We generally want to avoid an app having to do so, but provide escape hatches when necessary. `DOTNET_HOME` and app-local runtimes are the ways we are doing that. We may still decide we want `DOTNET_HOME` to point to the root of a directory containing versions and identify a specific version in that folder, but there is a lot of complexity there to manage and that complexity has to live in a native host executable that is hard to service ;)